packages/flutter/lib/src/widgets/modal_barrier.dart

```diff
@@ -12,6 +16,102 @@ import 'gesture_detector.dart';
 import 'navigator.dart';
 import 'transitions.dart';
 
+// Recognizes tap down by any pointer button unconditionally. When it receives a
+// PointerDownEvent, it immediately claims victor of arena and calls
+// `onAnyTapDown` without any checks.
+//
+// It is used by ModalBarrier to detect any taps on the overlay.
+class _AnyTapGestureRecognizer extends PrimaryPointerGestureRecognizer {
+  _AnyTapGestureRecognizer({ Object debugOwner }) : super(debugOwner: debugOwner);
+
+  VoidCallback onAnyTapDown;
+
+  bool _sentTapDown = false;
+
+  @override
+  SemanticsGestureConfiguration get semanticsConfiguration {
+    return _semanticsConfiguration ??= SemanticsGestureConfiguration(
+      onTap: onAnyTapDown,
+    );
+  }
+  SemanticsGestureConfiguration _semanticsConfiguration;
+
+  @override
+  void addAllowedPointer(PointerDownEvent event) {
+    super.addAllowedPointer(event);
+    resolve(GestureDisposition.accepted);
+  }
+
+  @override
+  void handlePrimaryPointer(PointerEvent event) {
+    if (!_sentTapDown) {
+      if (onAnyTapDown != null)
+        onAnyTapDown();
+      _sentTapDown = true;
+    }
+  }
+
+  @override
+  void didStopTrackingLastPointer(int pointer) {
+    super.didStopTrackingLastPointer(pointer);
+    _sentTapDown = false;
+  }
+
+  @override
+  String get debugDescription => 'anyTap';
+}
+
+// A GestureDetector used by ModalBarrier. It only has one callback,
+// `onAnyTapDown`, which recognizes tap down unconditionally.
+class _ModalBarrierGestureDetector extends StatelessWidget {
+  const _ModalBarrierGestureDetector({
+    Key key,
+    this.child,
+    this.onAnyTapDown,
+    this.behavior,
+    this.excludeFromSemantics = false,
+  }) : assert(excludeFromSemantics != null),
+       super(key: key);
+
+  /// The widget below this widget in the tree.
+  /// See [RawGestureDetector.child].
+  final Widget child;
+
+  /// Immediately called when a pointer causes a tap down.
+  /// See [_AnyTapGestureRecognizer.onAnyTapDown].
+  final VoidCallback onAnyTapDown;
+
+  /// How this gesture detector should behave during hit testing.
+  /// See [RawGestureDetector.behavior].
+  final HitTestBehavior behavior;
+
+  /// Whether to exclude these gestures from the semantics tree.
+  /// See [RawGestureDetector.excludeFromSemantics].
+  final bool excludeFromSemantics;
+
+  @override
+  Widget build(BuildContext context) {
+    final Map<Type, GestureRecognizerFactory> gestures = <Type, GestureRecognizerFactory>{};
+
+    if (onAnyTapDown != null) {
+      gestures[_AnyTapGestureRecognizer] = GestureRecognizerFactoryWithHandlers<_AnyTapGestureRecognizer>(
+        () => _AnyTapGestureRecognizer(debugOwner: this),
+        (_AnyTapGestureRecognizer instance) {
+          instance
+            ..onAnyTapDown = onAnyTapDown;
+        },
+      );
+    }
+
+    return RawGestureDetector(
+      gestures: gestures,
+      behavior: behavior,
+      excludeFromSemantics: excludeFromSemantics,
+      child: child,
+    );
+  }
+}
+
```

Comment:
we can probably optimize this widget some more since it's private and we know the one call site. For example, it looks like we never set excludeFromSemantics so we could just omit that entirely here, and it looks like the callback is never null so we could just assume that and assert it. In fact the callback never changes, we could just inline it right here so we don't have to pass a closure around.

Reply:
we could maybe even make the `gestures` map const.

Reply:
I've tried to optimize it. Unfortunately `gestures` can't be const since functions are involved.